File: python/src/aoc_2024_10.py
"""Advent of Code 2024 - Day 10."""

from sys import stdin

from coord import Coord, Extent


class Topo:
    """Topographic map."""

    def __init__(self, lines: list[str]) -> None:
        self.extent = Extent(len(lines[0].strip()), len(lines))
        self.h: list[list[int]] = [[int(ch) for ch in line.strip()] for line in lines]
        self.trail_heads: set[Coord] = {crd for crd in self.extent.upto() if self.ht(crd) == 0}

    def ht(self, crd: Coord) -> int:
        """Return height at given coordinate."""
        return self.h[crd.y][crd.x]

    def print(self) -> None:
        """Print heights for debugging."""
        for line in self.h:
            for h in line:
                print(h, end="")
            print()

    def count_trails(self, start: Coord) -> int:
        """Return number of 9-height positions reachable.

        Uses bfs.
        """
        visited: set[Coord] = set()
        frontier: set[Coord] = {start}
        count: int = 0
        while frontier:
            current = frontier.pop()
            visited.add(current)
            current_ht = self.ht(current)
            if current_ht == 9:
                count += 1
            else:
                for adjacent in set(current.adjacents(self.extent)) - visited - frontier:
                    if self.ht(adjacent) == current_ht + 1:
                        frontier.add(adjacent)
        return count

    def rating(self, start: Coord) -> int:
        """Return number of distinct hiking trails from start."""
        current: Coord = start
        rating: int = 0  # Number of paths found so far
        while True:
            current_height = self.ht(current)
            if current_height == 9:
                return rating + 1  # Found one path
            exits = [
                crd for crd in current.adjacents(self.extent) if self.ht(crd) == current_height + 1
            ]
            if not exits:
                return rating  # Reached a dead-end
            rating += sum(self.rating(x) for x in exits[1:])  # Check side-branches
            current = exits[0]  # And keep searching


if __name__ == "__main__":
    topo = Topo(stdin.readlines())
    print(sum(topo.count_trails(t) for t in topo.trail_heads))
    print(sum(topo.rating(t) for t in topo.trail_heads))

File: rust/src/bin/aoc_2024_10.rs
// Advent of Code 2024 - Day 10.

use aoc_rust::{ucoord, UCoord};
use grid::Grid;
use std::collections::HashSet;
use std::io;

fn read_topo_map() -> (Grid<u32>, HashSet<UCoord>) {
    let mut h: Grid<u32> = Grid::new(0, 0);
    for line in io::stdin().lines() {
        h.push_row(
            line.unwrap()
                .trim()
                .chars()
                .map(|ch| ch.to_digit(10).unwrap())
                .collect(),
        );
    }
    let trail_heads: HashSet<UCoord> = h
        .indexed_iter()
        .filter(|(_, h)| **h == 0)
        .map(|((row, col), _)| ucoord(row, col))
        .collect();

    (h, trail_heads)
}

fn adjacents(c: UCoord, (rows, cols): (usize, usize)) -> Vec<UCoord> {
    let mut v = Vec::with_capacity(4);
    if c.row > 0 {
        v.push(ucoord(c.row - 1, c.col));
    }
    if c.col > 0 {
        v.push(ucoord(c.row, c.col - 1));
    }
    if c.row < rows - 1 {
        v.push(ucoord(c.row + 1, c.col));
    }
    if c.col < cols - 1 {
        v.push(ucoord(c.row, c.col + 1));
    }
    v
}

fn count_trails(h: &Grid<u32>, start: UCoord) -> usize {
    let mut frontier: HashSet<UCoord> = HashSet::new();
    frontier.insert(start);
    let mut reached: HashSet<UCoord> = HashSet::new();
    while !frontier.is_empty() {
        let current = frontier.iter().next().copied().unwrap();
        frontier.remove(&current);
        let current_height = h[(current.row, current.col)];
        if current_height == 9 {
            reached.insert(current);
        } else {
            for a in adjacents(current, h.size()) {
                if !frontier.contains(&a) && h[(a.row, a.col)] == current_height + 1 {
                    frontier.insert(a);
                }
            }
        }
    }
    reached.len()
}

fn rating(h: &Grid<u32>, start: UCoord) -> usize {
    let mut current: UCoord = start;
    let mut paths: usize = 0;
    loop {
        let current_height = h[(current.row, current.col)];
        if current_height == 9 {
            return paths + 1;
        }
        let adjs = adjacents(current, h.size());
        let mut exits = adjs
            .iter()
            .filter(|c| h[(c.row, c.col)] == current_height + 1);
        if let Some(first_exit) = exits.next() {
            paths += exits.map(|x| rating(h, *x)).sum::<usize>();
            current = *first_exit;
        } else {
            return paths;
        }
    }
}

fn main() {
    let (h, trail_heads) = read_topo_map();
    let a: usize = trail_heads.iter().map(|t| count_trails(&h, *t)).sum();
    let b: usize = trail_heads.iter().map(|t| rating(&h, *t)).sum();
    print!("{a}\n{b}\n");
}
